File: c++/fibi.cpp
#include<bits/stdc++.h>
using namespace std;

//recursive
int fibi(int n){
    if(n<=1) return n;
    return fibi(n-1) + fibi(n-2);
}

//memorization
int fibi(int n, vector<int> &dp){
    if(n<=1) return n;
    if(dp.at(n)!=-1) return dp.at(n);
    return dp.at(n) = fibi(n-1, dp) + fibi(n-2, dp);
}

//tabulation
int fibiTabulation(int n){
    vector<int> dp(n+1,-1);
    dp.at(0) = 0;
    dp.at(1) = 1;
    for(int i=2; i<=n; i++){
        dp.at(i) = dp.at(i-1) + dp.at(i-2);
    }
    return dp.at(n);
}

//space complexity
int fibiSpaceComplexity(int n){
    int prev2 = 0;
    int prev = 1;
    for(int i=2; i<=n; i++){
        int cur = prev + prev2;
        prev2 = prev;
        prev = cur;
    }
    return prev;
}

int main(){
    //0,1,1,2,3,5,8,13
    //0,1,2,3,4,5,6, 7
    int n;
    cin>>n;
    vector<int> dp(n+1,-1);
    cout<<fibi(n)<<endl;
    cout<<fibi(n,dp)<<endl;
    cout<<fibiTabulation(n)<<endl;
    cout<<fibiSpaceComplexity(n)<<endl;
    return 0;
}

File: java/Fibi.java
import java.util.Scanner;

import javax.rmi.ssl.SslRMIClientSocketFactory;

import java.util.Arrays;

public class Fibi {
    static int recursive(int n){
        if(n<=1) return n;
        return recursive(n-1) + recursive(n-2);
    }

    static int memorization(int n, int[] dp ){
        if(n<=1) return n;
        if(dp[n] != -1) return dp[n];
        return dp[n] = memorization(n-1, dp) + memorization(n-2, dp);
        
    }

    static int tabulation(int n){
        int dp[] = new int[n+1];
        dp[0] = 0; 
        dp[1] = 1;
        for(int i=2; i<=n;  i++){
            dp[i] = dp[i-1] + dp[i-2];
        }
        return dp[n];
    }

    static int space_complexity(int n){
        int prev2 = 0; 
        int prev = 1;
        for(int i=2; i<=n; i++){
            int cur = prev + prev2;
            prev2 = prev;
            prev = cur;
        }
        return prev;
    }

    public static void main(String args[]){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.close();

        int dp [] = new int[n+1];
        Arrays.fill(dp,-1);

        System.out.println(recursive(n));
        System.out.println(memorization(n, dp));
        System.out.println(tabulation(n));
        System.out.println(space_complexity(n));
    }
}
